GitCommands/Settings/Setting.cs

```diff
@@ -4,91 +4,131 @@
+
+            public SettingOf(SettingsPath settingsSource, string name, T defaultValue = default)
             {
-                T v = Value;
-                if (ValueIsEmpty(v))
-                {
-                    return DefaultValue;
-                }
-                else
-                {
-                    return v;
-                }
+                SettingsSource = settingsSource;
+                Name = name;
+                DefaultValue = defaultValue;
             }
-        }
 
-        public virtual bool ValueIsEmpty(T value)
-        {
-            return EqualityComparer<T>.Default.Equals(value, default);
-        }
+            public SettingsPath SettingsSource { get; }
 
-        public string FullPath => SettingsSource.PathFor(Name);
+            public string Name { get; }
 
-        private T GetValue(string name, T defaultValue = default)
-        {
-            return SettingsSource.GetValue(name, defaultValue, value =>
+            public T DefaultValue { get; }
+
+            public T ValueOrDefault
             {
-                switch (Type.GetTypeCode(typeof(T)))
+                get
                 {
-                    case TypeCode.String:
-                        return (T)(object)value;
-                    default:
-                        return JsonConvert
-                            .DeserializeObject<T>(value) ?? defaultValue;
+                    T v = Value;
+
+                    if (ValueIsEmpty(v))
+                    {
+                        return DefaultValue;
+                    }
+                    else
+                    {
+                        return v;
+                    }
                 }
-            });
-        }
+            }
 
-        private void SetValue(string name, T value)
-        {
-            SettingsSource.SetValue(name, value, value =>
+            public T Value
             {
-                switch (Type.GetTypeCode(typeof(T)))
+                get => GetValue(Name, DefaultValue);
+                set
                 {
-                    case TypeCode.String:
-                        var stringValue = (string)(object)value;
-
-                        return string.IsNullOrEmpty(stringValue)
-                            ? null
-                            : stringValue;
-                    default:
-                        return JsonConvert
-                            .SerializeObject(value);
+                    if (Value.Equals(value))
+                    {
+                        return;
+                    }
+
+                    SetValue(Name, value);
+                    Updated?.Invoke(this, EventArgs.Empty);
                 }
-            });
+            }
+
+            public string FullPath => SettingsSource.PathFor(Name);
+
+            public bool ValueIsEmpty(T value)
+            {
+                return EqualityComparer<T>.Default.Equals(value, default);
+            }
```

Comment:
This makes my head spin.

Are we checking that:
* the value is unset, or 
* the value is default, or
* the value is empty?

Reply:
That the value is equal to the default value for the given type.
In the existing implementation, for data types, some value, for nullable is null.
For the given context, for value is empty or not set.

Reply:
Changed public to private